Modules/_datetimemodule.c

```diff
@@ -2880,26 +2880,34 @@ date_fromisoformat(PyObject *cls, PyObject *dtstr) {
         return NULL;
     }
 
-    Py_ssize_t len;
+    Py_ssize_t len = PyUnicode_GET_LENGTH(dtstr);
+    if (len != 10) {
+        goto invalid_string_error;
+    }
 
-    const char * dt_ptr = PyUnicode_AsUTF8AndSize(dtstr, &len);
+    PyObject * bytes = PyUnicode_AsASCIIString(dtstr);
+    if (bytes == NULL) {
+        return NULL;
+    }
+    const char * p = PyBytes_AS_STRING(bytes);
```

Comment:
Actually I missed this. I think it may be better to do `PyUnicode_AsUTF8` to avoid creating an unnecessary intermediate `bytes` object. If there are non-ascii characters in the string the parse step will fail.

Reply:
But then it will still fail on surrogate Unicode code points. Which after additional reading it apparently should.

It is strange to me that there is a way to directly get a `char *` for UTF-encoded strings, but not for ASCII.

Indeed the overhead here is likely too high, but using `PyUnicode_AsUTF8` here feels like a hack... maybe there's no better option though.

Reply:
I think it's fine. ASCII is a subset of UTF-8. The *actual* requirements just happen to include no non-ASCII (and thus no non-UTF8) characters in valid ISO 8601 strings, so it allows us to conveniently treat any encoding errors as a proxy for "this is not a valid ISO 8601 string".

There's no need to separately validate that it is both valid ASCII and a valid ISO 8601 string, because we only care if it's a valid ISO 8601 string. We should just do whatever is the most readable and performant way to figure that out (which, IMO, is "convert to UTF8, then try to parse it").